File: src/libqtemporarydir.cpp
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QTemporaryDir>
#include <qtemporarydir.h>
#include "libqtemporarydir.h"
#include "libqtemporarydir.hxx"

QTemporaryDir* QTemporaryDir_new() {
    return new QTemporaryDir();
}

QTemporaryDir* QTemporaryDir_new2(const libqt_string templateName) {
    QString templateName_QString = QString::fromUtf8(templateName.data, templateName.len);
    return new QTemporaryDir(templateName_QString);
}

void QTemporaryDir_Swap(QTemporaryDir* self, QTemporaryDir* other) {
    self->swap(*other);
}

bool QTemporaryDir_IsValid(const QTemporaryDir* self) {
    return self->isValid();
}

libqt_string QTemporaryDir_ErrorString(const QTemporaryDir* self) {
    QString _ret = self->errorString();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

bool QTemporaryDir_AutoRemove(const QTemporaryDir* self) {
    return self->autoRemove();
}

void QTemporaryDir_SetAutoRemove(QTemporaryDir* self, bool b) {
    self->setAutoRemove(b);
}

bool QTemporaryDir_Remove(QTemporaryDir* self) {
    return self->remove();
}

libqt_string QTemporaryDir_Path(const QTemporaryDir* self) {
    QString _ret = self->path();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

libqt_string QTemporaryDir_FilePath(const QTemporaryDir* self, const libqt_string fileName) {
    QString fileName_QString = QString::fromUtf8(fileName.data, fileName.len);
    QString _ret = self->filePath(fileName_QString);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

void QTemporaryDir_Delete(QTemporaryDir* self) {
    delete self;
}

File: src/libqtemporarydir.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://doc.qt.io/qt-6/qtemporarydir.html
pub const qtemporarydir = struct {
    /// New constructs a new QTemporaryDir object.
    ///
    ///
    pub fn New() QtC.QTemporaryDir {
        return qtc.QTemporaryDir_new();
    }

    /// New2 constructs a new QTemporaryDir object.
    ///
    /// ``` templateName: []const u8 ```
    pub fn New2(templateName: []const u8) QtC.QTemporaryDir {
        const templateName_str = qtc.struct_libqt_string{
            .len = templateName.len,
            .data = templateName.ptr,
        };

        return qtc.QTemporaryDir_new2(templateName_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qtemporarydir.html#swap)
    ///
    /// ``` self: QtC.QTemporaryDir, other: QtC.QTemporaryDir ```
    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QTemporaryDir_Swap(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qtemporarydir.html#isValid)
    ///
    /// ``` self: QtC.QTemporaryDir ```
    pub fn IsValid(self: ?*anyopaque) bool {
        return qtc.QTemporaryDir_IsValid(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qtemporarydir.html#errorString)
    ///
    /// ``` self: QtC.QTemporaryDir, allocator: std.mem.Allocator ```
    pub fn ErrorString(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QTemporaryDir_ErrorString(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qtemporarydir.ErrorString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qtemporarydir.html#autoRemove)
    ///
    /// ``` self: QtC.QTemporaryDir ```
    pub fn AutoRemove(self: ?*anyopaque) bool {
        return qtc.QTemporaryDir_AutoRemove(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qtemporarydir.html#setAutoRemove)
    ///
    /// ``` self: QtC.QTemporaryDir, b: bool ```
    pub fn SetAutoRemove(self: ?*anyopaque, b: bool) void {
        qtc.QTemporaryDir_SetAutoRemove(@ptrCast(self), b);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qtemporarydir.html#remove)
    ///
    /// ``` self: QtC.QTemporaryDir ```
    pub fn Remove(self: ?*anyopaque) bool {
        return qtc.QTemporaryDir_Remove(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qtemporarydir.html#path)
    ///
    /// ``` self: QtC.QTemporaryDir, allocator: std.mem.Allocator ```
    pub fn Path(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QTemporaryDir_Path(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qtemporarydir.Path: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qtemporarydir.html#filePath)
    ///
    /// ``` self: QtC.QTemporaryDir, fileName: []const u8, allocator: std.mem.Allocator ```
    pub fn FilePath(self: ?*anyopaque, fileName: []const u8, allocator: std.mem.Allocator) []const u8 {
        const fileName_str = qtc.struct_libqt_string{
            .len = fileName.len,
            .data = fileName.ptr,
        };
        const _str = qtc.QTemporaryDir_FilePath(@ptrCast(self), fileName_str);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qtemporarydir.FilePath: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qtemporarydir.html#dtor.QTemporaryDir)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QTemporaryDir ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QTemporaryDir_Delete(@ptrCast(self));
    }
};
